File: examples/java/com/intel/daal/examples/implicit_als/ImplAlsCSRDistr.java
/* ImplicitAlsCSRDistributed.java */
/*******************************************************************************
* Copyright 2014-2018 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
 //  Content:
 //     Java example of the implicit alternating least squares (ALS) algorithm in
 //     the distributed processing mode.
 //
 //     The program trains the implicit ALS model on a training data set.
 ////////////////////////////////////////////////////////////////////////////////
 */

/**
 * <a name="DAAL-EXAMPLE-JAVA-IMPLICITALSCSRDISTRIBUTED">
 * @example ImplAlsCSRDistr.java
 */

package com.intel.daal.examples.implicit_als;

import com.intel.daal.algorithms.implicit_als.PartialModel;
import com.intel.daal.algorithms.implicit_als.prediction.ratings.*;
import com.intel.daal.algorithms.implicit_als.training.*;
import com.intel.daal.algorithms.implicit_als.training.init.*;
import com.intel.daal.data_management.data.NumericTable;
import com.intel.daal.data_management.data.HomogenNumericTable;
import com.intel.daal.data_management.data.KeyValueDataCollection;
import com.intel.daal.examples.utils.Service;
import com.intel.daal.services.DaalContext;

class ImplAlsCSRDistr {
    /* Input data set parameters */
    private static final int      nNodes            = 4;
    private static final String[] dataset = {
            "../data/distributed/implicit_als_trans_csr_1.csv",
            "../data/distributed/implicit_als_trans_csr_2.csv",
            "../data/distributed/implicit_als_trans_csr_3.csv",
            "../data/distributed/implicit_als_trans_csr_4.csv" };

    private static final int    nUsers           = 46;           /* Full number of users */
    private static final int    nFactors         = 2;            /* Number of factors */
    private static final int    maxIterations    = 5;            /* Number of iterations in the implicit ALS training algorithm */

    private static NumericTable[] dataTable           = new NumericTable[nNodes];
    private static NumericTable[] dataTableTransposed = new NumericTable[nNodes];

    private static long[] usersPartition = { nNodes };

    private static NumericTable[] userOffsets = new NumericTable[nNodes];
    private static NumericTable[] itemOffsets = new NumericTable[nNodes];

    private static DistributedPartialResultStep1[] step1LocalResult = new DistributedPartialResultStep1[nNodes];
    private static NumericTable step2MasterResult;

    private static KeyValueDataCollection[] userStep3LocalInput = new KeyValueDataCollection[nNodes];
    private static KeyValueDataCollection[] itemStep3LocalInput = new KeyValueDataCollection[nNodes];

    private static KeyValueDataCollection[] step3LocalResult = new KeyValueDataCollection[nNodes];
    private static KeyValueDataCollection[] step4LocalInput  = new KeyValueDataCollection[nNodes];

    private static DistributedPartialResultStep4[] itemsPartialResultLocal = new DistributedPartialResultStep4[nNodes];
    private static DistributedPartialResultStep4[] usersPartialResultLocal = new DistributedPartialResultStep4[nNodes];

    private static NumericTable[][] predictedRatings = new NumericTable[nNodes][];

    private static DaalContext context = new DaalContext();

    public static void main(String[] args) throws java.io.FileNotFoundException, java.io.IOException {

        for (int iNode = 0; iNode < nNodes; iNode++) {
            readData(iNode);
            step4LocalInput[iNode] = new KeyValueDataCollection(context);
        }

        initializeModel();

        trainModel();

        testModel();

        printResults();

        context.dispose();
    }

    static KeyValueDataCollection initializeStep1Local(int iNode) {

        /* Create an algorithm object to perform the first step of the implicit ALS initialization algorithm */
        InitDistributedStep1Local initAlgorithm = new InitDistributedStep1Local(context, Float.class, InitMethod.fastCSR);
        initAlgorithm.parameter.setFullNUsers(nUsers);
        initAlgorithm.parameter.setNFactors(nFactors);
        initAlgorithm.parameter.setSeed(initAlgorithm.parameter.getSeed() + iNode);
        initAlgorithm.parameter.setPartition(new HomogenNumericTable(context, usersPartition, 1, usersPartition.length));

        /* Pass a training data set to the algorithm */
        initAlgorithm.input.set(InitInputId.data, dataTable[iNode]);

        /* Compute partial results of the first step on local nodes */
        InitPartialResult initPartialResult = initAlgorithm.compute();

        itemStep3LocalInput[iNode] = initPartialResult.get(InitPartialResultBaseId.outputOfInitForComputeStep3);
        userOffsets[iNode]         = initPartialResult.get(InitPartialResultBaseId.offsets, iNode);
        PartialModel partialModel  = initPartialResult.get(InitPartialResultId.partialModel);

        itemsPartialResultLocal[iNode] = new DistributedPartialResultStep4(context);
        itemsPartialResultLocal[iNode].set(DistributedPartialResultStep4Id.outputOfStep4ForStep1, partialModel);

        return initPartialResult.get(InitPartialResultCollectionId.outputOfStep1ForStep2);
    }

    static void initializeStep2Local(int iNode, KeyValueDataCollection initStep2LocalInput) {

        /* Create an algorithm object to perform the second step of the implicit ALS initialization algorithm */
        InitDistributedStep2Local initAlgorithm = new InitDistributedStep2Local(context, Float.class, InitMethod.fastCSR);

        initAlgorithm.input.set(InitStep2LocalInputId.inputOfStep2FromStep1, initStep2LocalInput);

        /* Compute partial results of the second step on local nodes */
        InitDistributedPartialResultStep2 initPartialResult = initAlgorithm.compute();

        dataTableTransposed[iNode] = initPartialResult.get(InitDistributedPartialResultStep2Id.transposedData);
        userStep3LocalInput[iNode] = initPartialResult.get(InitPartialResultBaseId.outputOfInitForComputeStep3);
        itemOffsets[iNode]         = initPartialResult.get(InitPartialResultBaseId.offsets, iNode);
    }

    static void initializeModel()
    {
        KeyValueDataCollection[] initStep1LocalResult = new KeyValueDataCollection[nNodes];
        for (int iNode = 0; iNode < nNodes; iNode++) {
            initStep1LocalResult[iNode] = initializeStep1Local(iNode);
        }

        /* Prepare input objects for the second step of the distributed initialization algorithm */
        KeyValueDataCollection[] initStep2LocalInput = new KeyValueDataCollection[nNodes];
        for (int iNode = 0; iNode < nNodes; iNode++) {
            initStep2LocalInput[iNode] = new KeyValueDataCollection(context);
            for (int jNode = 0; jNode < nNodes; jNode++) {
                initStep2LocalInput[iNode].set(jNode, initStep1LocalResult[jNode].get(iNode));
            }
        }

        for (int iNode = 0; iNode < nNodes; iNode++) {
            initializeStep2Local(iNode, initStep2LocalInput[iNode]);
        }
    }

    static void trainModel() {

        for (int iteration = 0; iteration < maxIterations; iteration++) {
            /* Update partial users factors */
            for (int iNode = 0; iNode < nNodes; iNode++) {
                computeStep1Local(iNode, itemsPartialResultLocal);
            }

            computeStep2Master();

            for (int iNode = 0; iNode < nNodes; iNode++) {
                computeStep3Local(iNode, itemOffsets[iNode], itemsPartialResultLocal, itemStep3LocalInput);
            }

            for (int iNode = 0; iNode < nNodes; iNode++) {
                computeStep4Local(iNode, dataTableTransposed, usersPartialResultLocal);
            }

            /* Update partial items factors */
            for (int iNode = 0; iNode < nNodes; iNode++) {
                computeStep1Local(iNode, usersPartialResultLocal);
            }

            computeStep2Master();

            for (int iNode = 0; iNode < nNodes; iNode++) {
                computeStep3Local(iNode, userOffsets[iNode], usersPartialResultLocal, userStep3LocalInput);
            }

            for (int iNode = 0; iNode < nNodes; iNode++) {
                computeStep4Local(iNode, dataTable, itemsPartialResultLocal);
            }
        }
    }

    static void computeStep1Local(int iNode, DistributedPartialResultStep4[] partialResultLocal) {
        /* Create an algorithm object to perform first step of the implicit ALS training algorithm on local-node data */
        DistributedStep1Local algorithm = new DistributedStep1Local(context, Float.class, TrainingMethod.fastCSR);
        algorithm.parameter.setNFactors(nFactors);

        /* Set input objects for the algorithm */
        algorithm.input.set(PartialModelInputId.partialModel,
                            partialResultLocal[iNode].get(DistributedPartialResultStep4Id.outputOfStep4ForStep1));

        /* Compute partial results of the first step on local nodes */
        step1LocalResult[iNode] = algorithm.compute();
    }

    static void computeStep2Master() {
        /* Create an algorithm object to perform second step of the implicit ALS training algorithm */
        DistributedStep2Master algorithm = new DistributedStep2Master(context, Float.class, TrainingMethod.fastCSR);
        algorithm.parameter.setNFactors(nFactors);

        /* Set the partial results of the first local step of distributed computations
           as input for the master-node algorithm */
        for (int i = 0; i < nNodes; i++)
        {
            algorithm.input.add(MasterInputId.inputOfStep2FromStep1, step1LocalResult[i]);
        }

        /* Compute a partial result on the master node from the partial results on local nodes */
        step2MasterResult = algorithm.compute().get(DistributedPartialResultStep2Id.outputOfStep2ForStep4);
    }

    static void computeStep3Local(int iNode, NumericTable offset, DistributedPartialResultStep4[] partialResultLocal,
                                  KeyValueDataCollection[] outBlocks) {
        /* Create an algorithm object to perform third step of the implicit ALS training algorithm on local-node data */
        DistributedStep3Local algorithm = new DistributedStep3Local(context, Float.class, TrainingMethod.fastCSR);
        algorithm.parameter.setNFactors(nFactors);

        /* Set input objects for the algorithm */
        algorithm.input.set(PartialModelInputId.partialModel,
                            partialResultLocal[iNode].get(DistributedPartialResultStep4Id.outputOfStep4ForStep3));
        algorithm.input.set(Step3LocalCollectionInputId.partialModelBlocksToNode, outBlocks[iNode]);
        algorithm.input.set(Step3LocalNumericTableInputId.offset, offset);

        /* Compute partial results of the third step on local nodes */
        DistributedPartialResultStep3 partialResult = algorithm.compute();

        /* Prepare input objects for the fourth step of the distributed algorithm */
        step3LocalResult[iNode] = partialResult.get(DistributedPartialResultStep3Id.outputOfStep3ForStep4);
        for (int i = 0; i < nNodes; i++) {
            step4LocalInput[i].set(iNode, step3LocalResult[iNode].get(i));
        }
    }

    static void computeStep4Local(int iNode, NumericTable[] dataTable, DistributedPartialResultStep4[] partialResultLocal) {
        /* Create an algorithm object to perform fourth step of the implicit ALS training algorithm on local-node data */
        DistributedStep4Local algorithm = new DistributedStep4Local(context, Float.class, TrainingMethod.fastCSR);
        algorithm.parameter.setNFactors(nFactors);

        /* Set input objects for the algorithm */
        algorithm.input.set(Step4LocalPartialModelsInputId.partialModels,        step4LocalInput[iNode]);
        algorithm.input.set(Step4LocalNumericTableInputId.partialData,           dataTable[iNode]);
        algorithm.input.set(Step4LocalNumericTableInputId.inputOfStep4FromStep2, step2MasterResult);

        /* Get the local implicit ALS partial models */
        partialResultLocal[iNode] = algorithm.compute();
    }

    private static void testModel() {

        for (int iNode = 0; iNode < nNodes; iNode++) {
            predictedRatings[iNode] = new NumericTable[nNodes];
            for (int jNode = 0; jNode < nNodes; jNode++) {
                /* Create an algorithm object to predict ratings based in the implicit ALS partial models */
                RatingsDistributed algorithm = new RatingsDistributed(context, Float.class, RatingsMethod.defaultDense);
                algorithm.parameter.setNFactors(nFactors);

                /* Set input objects for the algorithm */
                algorithm.input.set(RatingsPartialModelInputId.usersPartialModel,
                        usersPartialResultLocal[iNode].get(DistributedPartialResultStep4Id.outputOfStep4));
                algorithm.input.set(RatingsPartialModelInputId.itemsPartialModel,
                        itemsPartialResultLocal[jNode].get(DistributedPartialResultStep4Id.outputOfStep4));

                /* Predict ratings */
                algorithm.compute();

                /* Retrieve the algorithm results */
                predictedRatings[iNode][jNode] = algorithm.finalizeCompute().get(RatingsResultId.prediction);
            }
        }
    }

    private static void readData(int iNode) throws java.io.FileNotFoundException, java.io.IOException {

        /* Read dataset from a file and create a numeric table for storing the input data */
        dataTable[iNode] = Service.createSparseTable(context, dataset[iNode]);
    }

    private static void printResults() {
        for (int iNode = 0; iNode < nNodes; iNode++) {
            for (int jNode = 0; jNode < nNodes; jNode++) {
                System.out.println("Ratings for users block " + iNode + ", items block " + jNode + " :");
                Service.printALSRatings(userOffsets[iNode], itemOffsets[jNode], predictedRatings[iNode][jNode]);
            }
        }
    }
}

File: examples/daal/cpp/source/implicit_als/impl_als_csr_distr.cpp
/* file: impl_als_csr_distr.cpp */
/*******************************************************************************
* Copyright 2014 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of the implicit alternating least squares (ALS) algorithm in
!    the distributed processing mode.
!
!    The program trains the implicit ALS model on a training data set.
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-IMPLICIT_ALS_CSR_DISTRIBUTED"></a>
 * \example impl_als_csr_distr.cpp
 */

#include "daal.h"
#include "service.h"

using namespace daal;
using namespace daal::data_management;
using namespace daal::algorithms::implicit_als;

/* Input data set parameters */
const size_t nBlocks = 4;

const std::string trainDatasetFileNames[nBlocks] = {
    "../data/distributed/implicit_als_trans_csr_1.csv",
    "../data/distributed/implicit_als_trans_csr_2.csv",
    "../data/distributed/implicit_als_trans_csr_3.csv",
    "../data/distributed/implicit_als_trans_csr_4.csv"
};

static int usersPartition[] = { nBlocks };

NumericTablePtr userOffsets[nBlocks];
NumericTablePtr itemOffsets[nBlocks];

typedef float algorithmFPType; /* Algorithm floating-point type */

/* Algorithm parameters */
const size_t nUsers = 46; /* Full number of users */
const size_t nFactors = 2; /* Number of factors */
const size_t maxIterations = 5; /* Number of iterations in the implicit ALS training algorithm */

CSRNumericTablePtr dataTable[nBlocks];
CSRNumericTablePtr transposedDataTable[nBlocks];

NumericTablePtr predictedRatings[nBlocks][nBlocks];

KeyValueDataCollectionPtr userStep3LocalInput[nBlocks];
KeyValueDataCollectionPtr itemStep3LocalInput[nBlocks];

training::DistributedPartialResultStep4Ptr itemsPartialResultLocal[nBlocks];
training::DistributedPartialResultStep4Ptr usersPartialResultLocal[nBlocks];

void initializeModel();
void readData(size_t block);
void trainModel();
void testModel();
void printResults();

int main(int argc, char *argv[]) {
    checkArguments(argc,
                   argv,
                   4,
                   &trainDatasetFileNames[0],
                   &trainDatasetFileNames[1],
                   &trainDatasetFileNames[2],
                   &trainDatasetFileNames[3]);

    for (size_t i = 0; i < nBlocks; i++) {
        readData(i);
    }

    initializeModel();

    trainModel();

    testModel();

    printResults();

    return 0;
}

KeyValueDataCollectionPtr initializeStep1Local(size_t block) {
    /* Create an algorithm object to perform the first step of the implicit ALS initialization algorithm */
    training::init::Distributed<step1Local, algorithmFPType, training::init::fastCSR> initAlgorithm;

    /* Set parameters of the algorithm */
    initAlgorithm.parameter.fullNUsers = nUsers;
    initAlgorithm.parameter.nFactors = nFactors;
    initAlgorithm.parameter.seed += block;
    initAlgorithm.parameter.partition.reset(
        new HomogenNumericTable<int>((int *)usersPartition, 1, 1));

    /* Pass a training data set to the algorithm */
    initAlgorithm.input.set(training::init::data, dataTable[block]);

    /* Compute partial results of the first step on local nodes */
    initAlgorithm.compute();

    training::init::PartialResultPtr partialResult = initAlgorithm.getPartialResult();
    itemStep3LocalInput[block] = partialResult->get(training::init::outputOfInitForComputeStep3);
    userOffsets[block] = partialResult->get(training::init::offsets, block);
    PartialModelPtr partialModelLocal = partialResult->get(training::init::partialModel);

    itemsPartialResultLocal[block].reset(new training::DistributedPartialResultStep4());
    itemsPartialResultLocal[block]->set(training::outputOfStep4ForStep1, partialModelLocal);

    return partialResult->get(training::init::outputOfStep1ForStep2);
}

void initializeStep2Local(size_t block, KeyValueDataCollectionPtr initStep2LocalInput) {
    /* Create an algorithm object to perform the second step of the implicit ALS initialization algorithm */
    training::init::Distributed<step2Local, algorithmFPType, training::init::fastCSR> initAlgorithm;

    initAlgorithm.input.set(training::init::inputOfStep2FromStep1, initStep2LocalInput);

    /* Compute partial results of the second step on local nodes */
    initAlgorithm.compute();

    training::init::DistributedPartialResultStep2Ptr partialResult =
        initAlgorithm.getPartialResult();
    transposedDataTable[block] =
        CSRNumericTable::cast(partialResult->get(training::init::transposedData));
    userStep3LocalInput[block] = partialResult->get(training::init::outputOfInitForComputeStep3);
    itemOffsets[block] = partialResult->get(training::init::offsets, block);
}

void initializeModel() {
    KeyValueDataCollectionPtr initStep1LocalResult[nBlocks];
    for (size_t i = 0; i < nBlocks; i++) {
        initStep1LocalResult[i] = initializeStep1Local(i);
    }

    /* Prepare input objects for the second step of the distributed initialization algorithm */
    KeyValueDataCollectionPtr initStep2LocalInput[nBlocks];
    for (size_t i = 0; i < nBlocks; i++) {
        initStep2LocalInput[i].reset(new KeyValueDataCollection());
        for (size_t j = 0; j < nBlocks; j++) {
            (*initStep2LocalInput[i])[j] = (*initStep1LocalResult[j])[i];
        }
    }
    for (size_t i = 0; i < nBlocks; i++) {
        initializeStep2Local(i, initStep2LocalInput[i]);
    }
}

training::DistributedPartialResultStep1Ptr computeStep1Local(
    const training::DistributedPartialResultStep4Ptr &partialResultLocal) {
    /* Create an algorithm object to perform first step of the implicit ALS training algorithm on local-node data */
    training::Distributed<step1Local> algorithm;
    algorithm.parameter.nFactors = nFactors;

    /* Set input objects for the algorithm */
    algorithm.input.set(training::partialModel,
                        partialResultLocal->get(training::outputOfStep4ForStep1));

    /* Compute partial results of the first step on local nodes */
    algorithm.compute();

    /* Get the computed partial results */
    return algorithm.getPartialResult();
}

NumericTablePtr computeStep2Master(
    const training::DistributedPartialResultStep1Ptr *step1LocalResult) {
    /* Create an algorithm object to perform second step of the implicit ALS training algorithm */
    training::Distributed<step2Master> algorithm;
    algorithm.parameter.nFactors = nFactors;

    /* Set the partial results of the first local step of distributed computations
       as input for the master-node algorithm */
    for (size_t i = 0; i < nBlocks; i++) {
        algorithm.input.add(training::inputOfStep2FromStep1, step1LocalResult[i]);
    }

    /* Compute a partial result on the master node from the partial results on local nodes */
    algorithm.compute();

    /* Get the computed partial results */
    return algorithm.getPartialResult()->get(training::outputOfStep2ForStep4);
}

KeyValueDataCollectionPtr computeStep3Local(
    const NumericTablePtr &offsetTable,
    const training::DistributedPartialResultStep4Ptr &partialResultLocal,
    const KeyValueDataCollectionPtr &step3LocalInput) {
    /* Create an algorithm object to perform third step of the implicit ALS training algorithm on local-node data */
    training::Distributed<step3Local> algorithm;
    algorithm.parameter.nFactors = nFactors;

    /* Set input objects for the algorithm */
    algorithm.input.set(training::partialModel,
                        partialResultLocal->get(training::outputOfStep4ForStep3));
    algorithm.input.set(training::inputOfStep3FromInit, step3LocalInput);
    algorithm.input.set(training::offset, offsetTable);

    /* Compute partial results of the third step on local nodes */
    algorithm.compute();

    /* Get the computed partial results */
    return algorithm.getPartialResult()->get(training::outputOfStep3ForStep4);
}

training::DistributedPartialResultStep4Ptr computeStep4Local(
    const CSRNumericTablePtr &dataTable,
    const NumericTablePtr &step2MasterResult,
    const KeyValueDataCollectionPtr &step4LocalInput) {
    /* Create an algorithm object to perform fourth step of the implicit ALS training algorithm on local-node data */
    training::Distributed<step4Local> algorithm;
    algorithm.parameter.nFactors = nFactors;

    /* Set input objects for the algorithm */
    algorithm.input.set(training::partialModels, step4LocalInput);
    algorithm.input.set(training::partialData, dataTable);
    algorithm.input.set(training::inputOfStep4FromStep2, step2MasterResult);

    /* Build the implicit ALS partial model on the local node */
    algorithm.compute();

    /* Get the local implicit ALS partial models */
    return algorithm.getPartialResult();
}

void trainModel() {
    training::DistributedPartialResultStep1Ptr step1LocalResult[nBlocks];
    NumericTablePtr step2MasterResult;
    KeyValueDataCollectionPtr step3LocalResult[nBlocks];
    KeyValueDataCollectionPtr step4LocalInput[nBlocks];

    for (size_t i = 0; i < nBlocks; i++) {
        step4LocalInput[i].reset(new KeyValueDataCollection());
    }
    for (size_t iteration = 0; iteration < maxIterations; iteration++) {
        /* Update partial users factors */
        for (size_t i = 0; i < nBlocks; i++) {
            step1LocalResult[i] = computeStep1Local(itemsPartialResultLocal[i]);
        }
        step2MasterResult = computeStep2Master(step1LocalResult);

        for (size_t i = 0; i < nBlocks; i++) {
            step3LocalResult[i] = computeStep3Local(itemOffsets[i],
                                                    itemsPartialResultLocal[i],
                                                    itemStep3LocalInput[i]);
        }

        /* Prepare input objects for the fourth step of the distributed algorithm */
        for (size_t i = 0; i < nBlocks; i++) {
            for (size_t j = 0; j < nBlocks; j++) {
                (*step4LocalInput[i])[j] = (*step3LocalResult[j])[i];
            }
        }

        for (size_t i = 0; i < nBlocks; i++) {
            usersPartialResultLocal[i] =
                computeStep4Local(transposedDataTable[i], step2MasterResult, step4LocalInput[i]);
        }

        /* Update partial items factors */
        for (size_t i = 0; i < nBlocks; i++) {
            step1LocalResult[i] = computeStep1Local(usersPartialResultLocal[i]);
        }
        step2MasterResult = computeStep2Master(step1LocalResult);

        for (size_t i = 0; i < nBlocks; i++) {
            step3LocalResult[i] = computeStep3Local(userOffsets[i],
                                                    usersPartialResultLocal[i],
                                                    userStep3LocalInput[i]);
        }

        /* Prepare input objects for the fourth step of the distributed algorithm */
        for (size_t i = 0; i < nBlocks; i++) {
            for (size_t j = 0; j < nBlocks; j++) {
                (*step4LocalInput[i])[j] = (*step3LocalResult[j])[i];
            }
        }

        for (size_t i = 0; i < nBlocks; i++) {
            itemsPartialResultLocal[i] =
                computeStep4Local(dataTable[i], step2MasterResult, step4LocalInput[i]);
        }
    }
}

void testModel() {
    for (size_t i = 0; i < nBlocks; i++) {
        for (size_t j = 0; j < nBlocks; j++) {
            /* Create an algorithm object to predict ratings based in the implicit ALS partial models */
            prediction::ratings::Distributed<step1Local> algorithm;
            algorithm.parameter.nFactors = nFactors;

            /* Set input objects for the algorithm */
            algorithm.input.set(prediction::ratings::usersPartialModel,
                                usersPartialResultLocal[i]->get(training::outputOfStep4));
            algorithm.input.set(prediction::ratings::itemsPartialModel,
                                itemsPartialResultLocal[j]->get(training::outputOfStep4));

            /* Predict ratings */
            algorithm.compute();

            /* Retrieve the algorithm results */
            predictedRatings[i][j] = algorithm.getResult()->get(prediction::ratings::prediction);
        }
    }
}

void readData(size_t block) {
    /* Read trainDatasetFileName from a file and create a numeric table to store the input data */
    dataTable[block] = CSRNumericTablePtr(createSparseTable<float>(trainDatasetFileNames[block]));
}

void printResults() {
    for (size_t i = 0; i < nBlocks; i++) {
        for (size_t j = 0; j < nBlocks; j++) {
            std::cout << "Ratings for users block " << i << ", items block " << j << " :"
                      << std::endl;
            printALSRatings(userOffsets[i], itemOffsets[j], predictedRatings[i][j]);
        }
    }
}
